### PR TITLE
ms-python.python: fix #139813 and #137314 (lttng-ust and libstdc++ errors)

### LINKED ISSUE #137314: vscode-extensions.ms-python from unstable cannot be built due to too new lttng-ust
### Describe the bug
vscode-extensions.ms-python fails to build from unstable, as it requires liblttng-ust.so.0, but the latest version of the lttng-ust derivation (2.13.0) provides only liblttng-ust.so.1, and patchelf fails to to fix up the binaries from the vscode extension.

### Steps To Reproduce
1. attempt to build vscode-extensions.ms-python from unstable

### Expected behavior
Package builds

### Notify maintainers
@jraygauthier
@bjornfor 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.1, NixOS, 21.05.3001.12eb1d16ae3 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.15`
 - channels(root): `"nixos-unstable-21.11pre313614.09cd65b33c5, home-manager-21.05, nixos-21.05.3001.12eb1d16ae3, nixos-hardware"`
 - channels(dylan): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

### LINKED ISSUE #139813: libstdc++ error while installing vscode-extensions.ms-python.python
### Describe the bug
While trying to install this extension an error occurs. 

### Steps To Reproduce
`nix-shell -p vscode-extensions.ms-python.python`

### Expected behavior
Successful install

### Screenshots
N/A

### Additional context
Logs:
```
$ nix-shell -p vscode-extensions.ms-python.python
these derivations will be built:
  /nix/store/r6lq1zsf3y006113k3nh20yywazbq4zl-vscode-extension-ms-python-python-2021.5.829140558.drv
building '/nix/store/r6lq1zsf3y006113k3nh20yywazbq4zl-vscode-extension-ms-python-python-2021.5.829140558.drv'...
unpacking sources
unpacking source archive /nix/store/lxcn27jsgx7c5pcqw4m5lyz39caqj6hx-ms-python-python.zip
source root is extension
setting SOURCE_DATE_EPOCH to timestamp 1620673574 of file extension/README.md
patching sources
configuring
building
installing
post-installation fixup
patching script interpreter paths in /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558
checking for references to /build/ in /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558...
automatically fixing dependencies for ELF files
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_cython.cpython-38-x86_64-linux-gnu.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_cython.cpython-38-i386-linux-gnu.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_frame_eval/pydevd_frame_evaluator.cpython-38-i386-linux-gnu.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_frame_eval/pydevd_frame_evaluator.cpython-38-x86_64-linux-gnu.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/pydevd_attach_to_process/attach_linux_amd64.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/pydevd_attach_to_process/attach_linux_x86.so
  libstdc++.so.6 -> not found!
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libdbgshim.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/Microsoft.Python.LanguageServer
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/createdump
  libmscordaccore.so -> found: /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libmscordaccore.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30:/nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.Security.Cryptography.Native.OpenSsl.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libhostfxr.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.Native.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.Net.Http.Native.so
  libcurl.so.4 -> found: /nix/store/fzhgkh90jm1lgvm63rhbmvnbs3dl2657-curl-7.76.1/lib/libcurl.so.4
setting RPATH to: /nix/store/fzhgkh90jm1lgvm63rhbmvnbs3dl2657-curl-7.76.1/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libcoreclrtraceptprovider.so
  liblttng-ust.so.0 -> not found!
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.Net.Security.Native.so
  libgssapi_krb5.so.2 -> found: /nix/store/ga6avwlcsw00lfwwgpfrjw6s8rsshl7h-libkrb5-1.18/lib/libgssapi_krb5.so.2
setting RPATH to: /nix/store/ga6avwlcsw00lfwwgpfrjw6s8rsshl7h-libkrb5-1.18/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.IO.Compression.Native.so
  libz.so.1 -> found: /nix/store/43zy3j9a11icz158b336ahzf3bg7ykm1-zlib-1.2.11/lib/libz.so.1
setting RPATH to: /nix/store/43zy3j9a11icz158b336ahzf3bg7ykm1-zlib-1.2.11/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libmscordbi.so
  libmscordaccore.so -> found: /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libmscordaccore.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30:/nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libclrjit.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libcoreclr.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libmscordaccore.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/System.Globalization.Native.so
searching for dependencies of /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libhostpolicy.so
  libstdc++.so.6 -> found: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib/libstdc++.so.6
setting RPATH to: /nix/store/vyqnqbp3yczrzjx2byk6p680q3mzanyw-gcc-10.3.0-lib/lib
autoPatchelfHook could not satisfy dependency liblttng-ust.so.0 wanted by /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/languageServer.0.5.30/libcoreclrtraceptprovider.so
autoPatchelfHook could not satisfy dependency libstdc++.so.6 wanted by /nix/store/9f4z1942ssqcdbwmj8m0d5p8dkz8fpfn-vscode-extension-ms-python-python-2021.5.829140558/share/vscode/extensions/ms-python.python/pythonFiles/lib/python/debugpy/_vendored/pydevd/pydevd_attach_to_process/attach_linux_x86.so
Add the missing dependencies to the build inputs or set autoPatchelfIgnoreMissingDeps=true
builder for '/nix/store/r6lq1zsf3y006113k3nh20yywazbq4zl-vscode-extension-ms-python-python-2021.5.829140558.drv' failed with exit code 1
error: build of '/nix/store/r6lq1zsf3y006113k3nh20yywazbq4zl-vscode-extension-ms-python-python-2021.5.829140558.drv' failed
```

### Notify maintainers
@jraygauthier

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/qbflrvs01i9ygdf4dsmcxcacc22x1kg9-nix-info
copying path '/nix/store/qbflrvs01i9ygdf4dsmcxcacc22x1kg9-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.13.0-7614-generic, Pop!_OS, 20.04 LTS`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.14`
 - channels(scoder12): `"home-manager, nixpkgs-21.11pre318361.a3a23d9599b"`
 - nixpkgs: `/home/scoder12/.nix-defexpr/channels/nixpkgs`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
- vscode-extensions.ms-python.python
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by jraygauthier**
###### Motivation for this change

 -  Fix #139813
 -  Fix #137314

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Tested the extension on an actual python project launching `code` from the following environment:

```bash
$ nix-shell -I nixpkgs=. -p 'with import <nixpkgs> {}; vscode-with-extensions.override{ vscodeExtensions = with vscode-extensions; [ ms-python.python ms-toolsai.jupyter ]; }'
```

Was able to debug, lint, type-check and use auto completion successfully. Seems to work fine.

### CODE REVIEW

File: pkgs/misc/vscode-extensions/python/default.nix [hunk 1]
```diff
@@ -75,6 +75,21 @@ in vscode-utils.buildVscodeMarketplaceExtension rec {
     # Patch `packages.json` so that nix's *ctags* is used as default value for `python.workspaceSymbols.ctagsPath`.
     substituteInPlace "./package.json" \
       --replace "\"default\": \"ctags\"" "\"default\": \"${ctagsDefaultsTo}\""
+
+
+    # Similar cleanup to what's done in the `debugpy` python package.
+    # This prevent our autopatchelf logic to bark on unsupported binaries (`attach_x86.so`
+    # was problematic) but also should make our derivation less heavy.
+    (
+      set -x
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/python/default.nix:84]:
```suggestion
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -14507,6 +14507,16 @@ with pkgs;
 
   lttng-ust = callPackage ../development/tools/misc/lttng-ust { };
 
+  lttng-ust-2-10 = (lttng-ust.override {
+    liburcu = liburcu-0-12;
+  }).overrideAttrs (oldAttrs: rec {
+    version = "2.10.5";
+    src = fetchurl {
+      url = "https://lttng.org/files/lttng-ust/${oldAttrs.pname}-${version}.tar.bz2";
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:14515]:
```suggestion
      url = "https://lttng.org/files/lttng-ust/lttng-ust-${version}.tar.bz2";
```

Comment (jraygauthier) [reply pkgs/top-level/all-packages.nix:14515]: Done. Included in `aa62a2f7926e01d534f0dfcd9e7701cb6721ea7a`.

File: pkgs/misc/vscode-extensions/python/default.nix (hunk 1 above)
Comment (jraygauthier) [reply pkgs/misc/vscode-extensions/python/default.nix:84]: Done. Included in `95e30ee27ee8c18c659758d6b43b6816ada8a60f`.

**Comment by austinbutler** [comment]
Result of `nixpkgs-review pr 140564` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>liburcu-0-12</li>
    <li>lttng-ust-2-10</li>
    <li>vscode-extensions.ms-python.python</li>
  </ul>
</details>

**Review by austinbutler — APPROVED** [review]
Seems to work fine.

**Pull request merged**